### PR TITLE
Error variants

### PR DESCRIPTION
**Pull request by enetsee**
This PR introduces error variants for semantic and syntactic errors. 

The main change is to `Semantic_check.ml` which now uses monadic validation to accumulate semantic errors rather than using exceptions. The top level function `semantic_check_program` now returns a `Result.t` type indicating success or failure.

The top-level parser function `parse_file` and `parse_string` have also been changed to emit a result type but the lexer can still generate exceptions and generates warnings as a side effect.

The `stanc` executable has also been changed to accommodate the changes.

The PR leaves us kind of halfway between a pure approach and a side-effecting one. In particular: 

- The lexer still generates exceptions on failure
- The lexer still uses side-effecting functions for deprecation warnings
- The semantic check uses reference cells to manage front-end options
- The semantic check uses side-effecting functions for modifying the symbol table

The third of these also means that we can't yet hide the frontend implementation using the signature in `Middle.Frontend_intf` since `stanc` directly sets mutable variables in frontend.

### CODE REVIEW

File: src/frontend/Errors.ml [hunk 1]
```diff
@@ -5,30 +5,30 @@ open Middle
...
-exception SemanticError of (string * location_span)
+exception SemanticError of (string * Mir.location_span)
 
 (** Exception [FatalError [msg]] indicates an error that should never happen with message
     [msg]. *)
 exception FatalError of string
 
 (* A semantic error reported by the toplevel *)
-let semantic_error ~loc msg = raise (SemanticError (msg, loc))
+(* let semantic_error ~loc msg = raise (SemanticError (msg, loc)) *)
```
Comment (VMatthijs) [inline src/frontend/Errors.ml]: Can this be deleted?

File: src/frontend/Errors.mli [hunk 2]
```diff
@@ -18,8 +18,8 @@ exception SemanticError of (string * location_span)
     so we can trace their origin. *)
 exception FatalError of string
 
-val semantic_error : loc:location_span -> string -> 'a
-(** Throw a semantic error reported by the toplevel *)
+(* val semantic_error : loc:location_span -> string -> 'a
```
Comment (VMatthijs) [inline src/frontend/Errors.mli]: Can this be deleted?

**Comment by enetsee** [comment]
I've got a bug caused by not adding an id to the symbol table - I'll update when it's fixed.

File: src/frontend/Parse.ml [hunk 3]
```diff
@@ -2,7 +2,15 @@
...
 open Core_kernel
-open Errors
+open Middle
+
+type syntax_error = Parsing of string * Mir.location_span
+
+let syntax_error_message = function Parsing (msg, _) -> msg
+let syntax_error_location = function Parsing (_, loc) -> loc
+
+(* TODO :render messages along with location! *)
```
Comment (VMatthijs) [inline src/frontend/Parse.ml]: Can you please still do this?

File: src/frontend/Parse.ml [hunk 4]
```diff
@@ -208,6 +221,8 @@ let%expect_test "parse indices, two different colons" =
   let ast =
     parse_string Parser.Incremental.program
       "model { matrix[5, 5] x; print(x[2 - 3 ? 3 : 4 : 2]); }"
+    |> Result.map_error ~f:render_syntax_error
+    |> Result.ok_or_failwith
```
Comment (VMatthijs) [inline src/frontend/Parse.ml]: Does it make sense to make these handlers already part of the parsing function? I just keep seeing this combination repeated.

Comment (enetsee) [reply src/frontend/Parse.ml]: Yes, the idea is to hide of this from stanc by using the `Frontend.S` module signature. I can't do that right now (reason outlined above) so tried to keep things as-is until we make the change.

File: src/frontend/Semantic_error.ml [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+    | IllTypedBinaryOperator of operator * unsizedtype * unsizedtype
+    | IllTypedPrefixOperator of operator * unsizedtype
+    | IllTypedPostfixOperator of operator * unsizedtype
+    | NotIndexable of unsizedtype
+
+  (* let assignmentoperator_to_stan_math_fn = function
+    | Plus -> Some "assign_add"
+    | Minus -> Some "assign_subtract"
+    | Times -> Some "assign_multiply"
+    | Divide -> Some "assign_divide"
```
Comment (VMatthijs) [inline src/frontend/Semantic_error.ml]: What are these definitions doing here? Can they be deleted?

File: src/middle/Backend_intf.ml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+open Mir
+
+module type S = sig
```
Comment (VMatthijs) [inline src/middle/Backend_intf.ml]: Is this at all related to the PR?

File: src/middle/Compiler.ml [hunk 7]
```diff
@@ -1,124 +1,32 @@
...
+include Core_kernel
+
+module Make
+    (F : Frontend_intf.S)
+    (O : Optimization_intf.S)
+    (B : Backend_intf.S) : S with type frontend_error := F.frontend_error =
+struct
+  type compiler_opts =
+    { frontend_opts: F.frontend_opts
+    ; optimization_opts: O.optimization_opts
```
Comment (VMatthijs) [inline src/middle/Compiler.ml]: Is this at all related to the PR?

File: src/middle/Compiler_intf.ml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  type compiler_opts_error
+  type compiler_opts
+
+  val default_compiler_opts : compiler_opts
+
+  val compiler_opts_of_string :
+    string -> (compiler_opts, compiler_opts_error list) result
+
+  val compile_from_file :
+    opts:compiler_opts -> file:string -> (string, frontend_error) result
```
Comment (VMatthijs) [inline src/middle/Compiler_intf.ml]: Is this at all related to the PR? Also, just wondering: does it matter whether this is a .ml or .mli file?

File: src/middle/Compiler.ml (hunk 7 above)
Comment (enetsee) [reply src/middle/Compiler.ml]: No - I tidied it up before I realized I wasn't going to be able to use the signatures yet.

File: src/middle/Middle.ml [hunk 9]
```diff
@@ -1,12 +1,37 @@
 include Mir
 open Core_kernel
+module Mir = Mir
```
Comment (VMatthijs) [inline src/middle/Middle.ml]: What does this do?

File: src/middle/Compiler_intf.ml (hunk 8 above)
Comment (enetsee) [reply src/middle/Compiler_intf.ml]: Same as above regarding; it's in an .ml file since we want to expose the module signatures and functor from middle so that it can be used to either write new frontends/backends or construct the core compiler with them

File: src/middle/Middle.ml (hunk 9 above)
Comment (enetsee) [reply src/middle/Middle.ml]: Before this change, Mir was being _included_ in Middle.  With this approach, Mir is exposed as a module  which I think is a little nicer since `Middle` contains more than just `Mir`.

As a consequence, you can now open `Middle` and not have all the `Mir` types in scope.

File: src/middle/Backend_intf.ml (hunk 6 above)
Comment (enetsee) [reply src/middle/Backend_intf.ml]: No - I tidied it up before I realized I wasn't going to be able to use the signatures yet.

File: src/stanc/stanc.ml [hunk 10]
```diff
@@ -79,25 +79,43 @@ let add_file filename =
 (** ad directives from the given file. *)
 let use_file filename =
   let ast =
-    try Parse.parse_file Parser.Incremental.program filename
-    with Errors.SyntaxError err ->
-      Errors.report_syntax_error err ;
+    try 
+      match Parse.parse_file Parser.Incremental.program filename with
+      | Result.Ok ast -> ast
```
Comment (VMatthijs) [inline src/stanc/stanc.ml]: So does it say OK even if there is a lexing error, for instance?

File: src/stanc/stanc.ml [hunk 11]
```diff
@@ -79,25 +79,43 @@ let add_file filename =
...
+      | Result.Error err ->
+          let loc = Parse.syntax_error_location err
+          and msg = Parse.syntax_error_message err in
+          Errors.report_parsing_error (msg, loc) ;
+          exit 1
+    with Errors.SyntaxError err ->
```
Comment (VMatthijs) [inline src/stanc/stanc.ml]: Is this only to catch the syntax errors that are not parsing errors? If so, should the parsing errors then be removed from the syntax errors? I feel like the syntax errors and their handling could be simplified further.

File: src/stanc/stanc.ml (hunk 10 above)
Comment (enetsee) [reply src/stanc/stanc.ml]: No, lexing errors still raise exceptions at the moment which is outside the context of validation.

File: src/stanc/stanc.ml (hunk 11 above)
Comment (enetsee) [reply src/stanc/stanc.ml]: Yes, exactly. As I mentioned in the first message, we're sort of halfway between a monadic and exception based approach right now.

Comment (enetsee) [reply src/stanc/stanc.ml]: And I agree about simplifying further. Ideally we want no exceptions coming from `frontend` - keep the core of the compiler pure and do all the IO in the `stanc` wrapper.

File: src/middle/Validation_intf.ml [hunk 12]
```diff
@@ -0,0 +1,42 @@
+module type Infix = sig
+  type 'a t
+
+  val ( >>= ) : 'a t -> ('a -> 'b t) -> 'b t
```
Comment (VMatthijs) [inline src/middle/Validation_intf.ml]: There's part of me that really likes abstracting out this sort of validation monad because obviously that is what is going on with this sort of error accumulation. Another part of me wonders whether it's slightly overengineered and would make the semantic check harder to understand for new developers and harder to maintain.
Let me think about it a bit longer. I think I mostly like it.

File: src/frontend/Semantic_error.ml [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+    | DuplicateArgNames ->
+        Fmt.pf ppf "All function arguments must have distinct identifiers."
+    | IncompatibleReturnType ->
+        Fmt.pf ppf
+          "Function bodies must contain a return statement of correct type in \
+           every branch."
+end
+
+type t =
+  | TypeError of location_span * TypeError.t
```
Comment (VMatthijs) [inline src/frontend/Semantic_error.ml]: This is super slick! :-) I should've done that immediately when I wrote it. The problem was that I had no idea what all the error conditions would be and how they should be organized when I started writing the semantic check.

File: src/middle/Validation_intf.ml (hunk 12 above)
Comment (enetsee) [reply src/middle/Validation_intf.ml]: Yeah, I get where you're coming from. 

For type-checking / semantic-checking, what we actually want is a monad transfomer stack with `Reader` for the options, `Validation` to accumulate errors and `State` to carry the environment around.

Clearly this is going to be quite intimidating for most developers and probably impenetrable for someone with out an FP background. 

On the plus side, I this the approach will make something like a language server easier and should make the experience nicer for users i.e. we can report syntactic and semantic errors together rather than just the first.

File: src/frontend/Semantic_check.ml [hunk 14]
```diff
@@ -170,164 +104,67 @@ let probability_distribution_name_variants id =
...
-   |AnyReturnType, AnyReturnType ->
-      AnyReturnType
+      Validate.ok (Mir.ReturnType UReal)
+  | _, _ when rt1 = rt2 -> Validate.ok rt2
+  | _ -> Semantic_error.mismatched_return_types loc rt1 rt2 |> Validate.error
 
 let check_fresh_variable_basic id is_nullary_function =
-  (* No shadowing! *)
-  (* For some strange reason, Stan allows user declared identifiers that are
+  Validate.(
```
Comment (VMatthijs) [inline src/frontend/Semantic_check.ml]: Could you maybe explain what this line does? Validate is a functor, right? It's Validation.Make. So I would expect it to apply to be applied to a module which contains a single type. Here, it seems to be applied to a value of result type. Is that related to the dot after Validate? Could you help me parse this? Thanks!

Comment (enetsee) [reply src/frontend/Semantic_check.ml]: At the moment, `Validation` is written so that you construct it via `Validation.Make` passing in another module which has some type `t` which is the type of error you are going to accumulate.

In `semantic_check.ml`, I create a module `Validate` by passing `Semantic_error` into `Validation.Make` fixing the type of error at `Semantic_error.t`.

When I write `Validate.( ... )`, this `Validate` module is in scope inside the brackets. This is particularly helpful when using infix functions like `>>=`; without it I would have to qualify everything.

Comment (VMatthijs) [reply src/frontend/Semantic_check.ml]: Oh gotcha! So would this be equivalent to writing a 
` let open Validate` first?

Comment (enetsee) [reply src/frontend/Semantic_check.ml]: Yes - there are two ways of writing local open. As well as being more lightweight, I always think this syntax mirrors what I'm doing in my head.

Comment (VMatthijs) [reply src/frontend/Semantic_check.ml]: Cool, thanks!

File: src/middle/Validation_intf.ml (hunk 12 above)
Comment (enetsee) [reply src/middle/Validation_intf.ml]: With all that being said, I think we can actually make the API to the `SemanticCheck` monad quite nice so that beginners can almost treat it like a DSL.

Comment (VMatthijs) [reply src/middle/Validation_intf.ml]: OK. I've thought about it more and don't see a way around it. I think this is more or less the most slick and least intimidating way of achieving what we want.

File: src/middle/Compiler_intf.ml (hunk 8 above)
Comment (VMatthijs) [reply src/middle/Compiler_intf.ml]: Could you not do that if it's in a .mli file? I guess I still don't fully understand when you put signatures in a .mli file and when in a .ml file.

File: src/middle/Middle.ml (hunk 9 above)
Comment (VMatthijs) [reply src/middle/Middle.ml]: Gotcha, thanks!

File: src/stanc/stanc.ml (hunk 11 above)
Comment (VMatthijs) [reply src/stanc/stanc.ml]: Agreed about throwing the exceptions as far to the outside as possible. That's the only way I can imagine accumulating more than one exception. We'd still have to think about how to continue parsing after an error, but presumably we could figure that out based on what Merlin does.

**Comment by VMatthijs** [comment]
I think this is great. Just let me know when it passes all tests and I'll have another look!

File: src/middle/Mir_intf.ml [hunk 15]
```diff
@@ -0,0 +1,200 @@
+open Core_kernel
```
Comment (seantalts) [inline src/middle/Mir_intf.ml]: You mentioned this at the meeting but I had something totally different in mind. What is the idea behind this Mir_intf abstraction? We aren't going to have multiple MIRs supported...

Comment (enetsee) [reply src/middle/Mir_intf.ml]: The idea here isn't to hide the MIR implementation but rather to have a `Mir` module exposed. 

The way it is at the moment means that all the MIR types are in scope as soon as you open `Middle`; since `Middle` now contains a bunch of other stuff I wanted to scope them.

Having the signature contained in `Mir_intf.mli` means I can include it in `Middle.mli` to get the nested module.

Comment (seantalts) [reply src/middle/Mir_intf.ml]: I think if we want to partition off the stuff in Middle I would go the other way - put the e.g. pretty printer in its own module, etc. I think the most important part of the Middle (and its main API) is the MIR and that should be front and center.

Comment (seantalts) [reply src/middle/Mir_intf.ml]: Also just one meta comment - it seems like this PR could have been broken into a few smaller ones to be easier to handle - for example, one could have just dealt with the changes to error variants and semantic check, another could have refactored the Middle, another could have done the refactorings necessary to get the different compiler ends into shape to fit into the compiler module interface you wrote. It's probably too late now for this PR (definitely don't want to make you do a ton of work to try to separate them out or redo them), but just for future PR harmony :)

Comment (enetsee) [reply src/middle/Mir_intf.ml]: Yeah - I did a bunch of stuff to `Middle` thinking I had done enough with the frontend to get the abstraction working; turns out I hadn't. It won't be a lot of work to revert `Middle` and I agree that this PR is overreaching a little!

**Comment by enetsee** [comment]
I've created a new file in middle `Mir_pretty_printer.ml` during the merge from master. This was to avoid a cycle between `Stan_math_signatures.ml` (which was moved to `Middle` on master) and `Middle.ml`.

**Comment by enetsee** [comment]
@VMatthijs  @seantalts this is ready for review now :fireworks:

**Comment by seantalts** [comment]
Okay cool! It might take me a little while, 3000 lines is quite a lot! Minor, but it seems like maybe we could just take out all the Compiler related stuff - esp because I'm seeing the _intf pattern that we're still discussing (if we can use the "module type of" pattern I think we should do that instead if it's equivalent).

File: src/frontend/Ast_to_Mir.ml [hunk 16]
```diff
@@ -3,8 +3,10 @@ open Middle
 
 (* XXX fix exn *)
 let unwrap_return_exn = function
-  | Some (ReturnType ut) -> ut
-  | x -> raise_s [%message "Unexpected return type " (x : returntype option)]
+  | Some (Middle.ReturnType ut) -> ut
+  | x ->
+      raise_s
+        [%message "Unexpected return type " (x : Middle.returntype option)]
```
Comment (seantalts) [inline src/frontend/Ast_to_Mir.ml]: We opened Middle at the top - we shouldn't need to change lines like these, probably an artifact from modularizing the Mir type.

File: src/frontend/Ast_to_Mir.ml [hunk 17]
```diff
@@ -136,14 +138,48 @@ let trans_printables mloc (ps : Ast.typed_expression Ast.printable list) =
       | Ast.PExpr e -> trans_expr e)
     ps
 
+let unsizedtype_of_indexed ut
```
Comment (seantalts) [inline src/frontend/Ast_to_Mir.ml]: What motivated not using the Semantic check function anymore?

File: src/frontend/Errors.ml [hunk 18]
```diff
@@ -1,34 +1,30 @@
 (** Setup of our compiler errors *)
 
 open Core_kernel
-open Middle
```
Comment (seantalts) [inline src/frontend/Errors.ml]: What's bad about opening Middle? I don't see any ambiguity and editors are quick to resolve to the right place with goto-definition.

File: src/frontend/Errors.ml [hunk 19]
```diff
@@ -1,34 +1,30 @@
...
-(* A semantic error reported by the toplevel *)
-let semantic_error ~loc msg = raise (SemanticError (msg, loc))
-
 (* A fatal error reported by the toplevel *)
 let fatal_error ?(msg = "") _ =
   raise (FatalError ("This should never happen. Please file a bug. " ^ msg))
 
 (** Parse a string into a location *)
-let rec parse_location str =
+let rec parse_location str : Middle.location =
```
Comment (seantalts) [inline src/frontend/Errors.ml]: Why add type signatures?

Comment (enetsee) [reply src/frontend/Errors.ml]: I added when type inference was going a bit nuts; should be ok to remove

File: src/frontend/Errors.ml (hunk 18 above)
Comment (enetsee) [reply src/frontend/Errors.ml]: Force of habit - I almost never open modules at the top level. I'll open and remove the qualifications

File: src/frontend/Ast_to_Mir.ml (hunk 17 above)
Comment (enetsee) [reply src/frontend/Ast_to_Mir.ml]: During the refactor, the function in `Semantic_check` was changed to return an `unsizedtype Validate.t` since there is a possibility that the indexed type is not semantically valid. 

In`Ast_to_Mir` we are only ever working with semantically correct ASTs so we should always be able to return an `unsizedtype`. Any deviation from this is genuinely exceptional behavior so we should use exceptions here to handle that case.

Whilst I could have reused the function inside `Semantic_check`, converted the `Validate.t` to a result and then raised an exception in the `Error` case, it would have meant exposing the `Validate` module outside of the semantic check module and violating DRY seemed a little nicer.

An alternative would be to expose a function `unsizedtype_of_indexed_exn` from `Semantic_check` although it would be a little odd in terms of the minimal API that module exposes. 

Happy to do the latter if you think it makes more sense?

File: src/middle/Middle.mli [hunk 20]
```diff
@@ -1,27 +1,46 @@
-include module type of Mir
 open Core_kernel
+include module type of Mir
+include module type of Mir_pretty_printer
+module Validation : Validation_intf.Validation
```
Comment (VMatthijs) [inline src/middle/Middle.mli]: Do we still need this `_intf` pattern here? Sean seemed to be able to make the `include module type of Foo` syntax work even in cases where there was a `Foo.mli` file. Perhaps it can't be made to work at all with functors?

File: src/frontend/Semantic_error.mli [hunk 21]
```diff
@@ -0,0 +1,79 @@
+open Middle
+
+type t
+
```
Comment (VMatthijs) [inline src/frontend/Semantic_error.mli]: This splitting out of the Semantic_error conditions is really great. :-)

File: src/frontend/Ast_to_Mir.ml (hunk 17 above)
Comment (VMatthijs) [reply src/frontend/Ast_to_Mir.ml]: Is there somewhere neutral that we can put it so that both files have access to it? It seems unnecessary to duplicate that function, particularly because I remember that its logic is a bit subtle. This would be my only real requested change before this can be merged.

File: src/middle/Validation.ml [hunk 22]
```diff
@@ -0,0 +1,55 @@
+include Validation_intf
```
Comment (VMatthijs) [inline src/middle/Validation.ml]: Actually, could you add a little bit of documentation to the Validation monad? A minimum would be a comment at the top to explain what it's supposed to do (e.g. accumulate errors).

File: src/frontend/Ast_to_Mir.ml (hunk 17 above)
Comment (enetsee) [reply src/frontend/Ast_to_Mir.ml]: Sure - I think the second option above is most sensible i.e. expose expose a function `unsizedtype_of_indexed_exn` which unsafely coverts the `unsizedtype Validate.t` into an `unsizedtype`.

File: src/middle/Middle.mli (hunk 20 above)
Comment (enetsee) [reply src/middle/Middle.mli]: I'll take a proper look at Sean's approach and see if it can be done.

File: src/middle/Validation.ml (hunk 22 above)
Comment (enetsee) [reply src/middle/Validation.ml]: Will do - good shout.

File: src/frontend/Parse.ml (hunk 3 above)
Comment (enetsee) [reply src/frontend/Parse.ml]: Done

File: src/frontend/Semantic_error.ml (hunk 5 above)
Comment (enetsee) [reply src/frontend/Semantic_error.ml]: Done

**Comment by VMatthijs** [comment]
Woohoo! Well done!
